Automatically grow the set's size

thass0 opened the pull request:
- Add a `new` function to create a set with an automatic size.
- Add a `set!` macro to initialize sets.
- Remove name ambiguity between capacity and maximum value

Closes #4 and #5.

thass0 committed:
Automatically grow the set's size

- Add a `new` function to create a set with an
  automatic size.
- Add a `set!` macro to initialize sets.
- Remove name ambiguity between capacity and maximum
  value (see issue #5)

Chriscbr commented on src/lib.rs:
```diff
@@ -43,41 +42,63 @@ use num_traits::Unsigned;
 /// A sparse set of integer values.
 pub struct SparseSet<T: PrimInt + Unsigned> {
-    max_value: T,
+    max_value: usize,
```
I wonder if we should rename this to `cap` or `capacity`

thass0 replied:
~~Storing `max_value` instead of `cap` is more efficient, because `max_value` is used to check if the given value lies within the current bounds on each call to `contains`, `insert`, etc.~~ The two can be converted into each other easily with `max_value = cap - 1`. ~~`cap` is used more rarely so it should be the one that's computed from `max_value`.~~ (Edit: this was more of a problem when `max_value` had the type `T`. Since it's `usize`, there is basically no need to convert, see commit [316817d](https://github.com/Chriscbr/thinset/pull/6/commits/316817d14ff1a77d100e0d350029216f08ecef62)).

thass0 replied:
Using `max_value` has one problem though: if `cap = 0`, then there is no `max_value`. As is, this cannot be expressed. Also, the calculation of `max_value` in `with_capacity` is wrong. 😅 Fixing that.

Chriscbr commented on src/lib.rs:
```diff
@@ -300,14 +412,66 @@ mod tests {
+        let mut s: SparseSet<u32> = SparseSet::new();
+    
+        // Check that inserting random values works.
+        for _ in 0..n_iters {
+            let x = rng.gen_range(0 .. 10000);
+           s.insert(x);
+           check.push(x);
+        }
```
nit
```suggestion
        for _ in 0..n_iters {
            let x = rng.gen_range(0 .. 10000);
            s.insert(x);
            check.push(x);
        }
```

Chriscbr commented on src/lib.rs:
````diff
@@ -158,6 +200,31 @@ impl<T: PrimInt + Unsigned> SparseSet<T> {
+///
+/// ```rust
+/// use thinset::{set, SparseSet};
+/// let mut set: SparseSet<u32> = set!(4, 32, 16, 24, 63);
````
Does this work with square brackets? I think that's the more conventional style, see e.g. https://doc.rust-lang.org/std/macro.vec.html
```suggestion
/// let mut set: SparseSet<u32> = set![4, 32, 16, 24, 63];
```

thass0 replied:
Yes, all kinds of brackets work. I think they can be chosen at the call site. I felt like using square brackets suggested that one is initializing an ordered structure.

Chriscbr reviewed (changes requested):
This looks great! Just a few nitpicks

thass0 committed:
Use square brackets with set! macro

thass0 committed:
Store capacity instead of maximum value

The overhead of storing the capacity is the
same as storing the maximum value. The advantage
of using the capacity is that capacities of 0 can
be expressed where no value can be stored at all
(at least not until the set is resized).

Chriscbr committed:
run cargo fmt
